docs/code_from_chapters/usecase_ml_code.rst

```diff
@@ -0,0 +1,840 @@
+       return image_list
+
+
+   def load_labels(data_frame, column_name):
+       label_list = data_frame[column_name].to_list()
+       return label_list
+
+
+   def preprocess(image):
+       resized = resize(image, (100, 100, 3))
+       reshaped = resized.reshape((1, 30000))
+       return reshaped
+
+
+   def load_data(data_path):
+       df = pd.read_csv(data_path)
+       labels = load_labels(data_frame=df, column_name="label")
+       raw_images = load_images(data_frame=df, column_name="filename")
+       processed_images = [preprocess(image) for image in raw_images]
+       data = np.concatenate(processed_images, axis=0)
+       return data, labels
+
+
+   def main(repo_path):
+       train_csv_path = repo_path / "data/train.csv"
+       train_data, labels = load_data(train_csv_path)
+       sgd = SGDClassifier(max_iter=10)
+       trained_model = sgd.fit(train_data, labels)
+       dump(trained_model, repo_path / "model.joblib")
+
+
+   if __name__ == "__main__":
+       repo_path = Path(__file__).parent.parent
+       main(repo_path)
+   EOT
+
+Let's save it::
+
+   datalad save -m "Add SGD classification script" code/train.py
+
+The next script loads the trained classifier from disk and evaluates it on the validation data.
+To evaluate the model performance, it calculates the accuracy of the prediction, i.e., the proportion of correctly labeled images, prints it to the terminal, and saves it into a json file in the superdataset::
+
+   cat << EOT > code/evaluate.py
+
+   #!/usr/bin/env python3
+
+   from joblib import load
+   import json
+   from pathlib import Path
+
+   from sklearn.metrics import accuracy_score
+
+   from train import load_data
+
+
+   def main(repo_path):
+       test_csv_path = repo_path / "data/test.csv"
+       test_data, labels = load_data(test_csv_path)
+       model = load(repo_path / "model.joblib")
+       predictions = model.predict(test_data)
+       accuracy = accuracy_score(labels, predictions)
+       metrics = {"accuracy": accuracy}
+       print(metrics)
+       accuracy_path = repo_path / "accuracy.json"
+       accuracy_path.write_text(json.dumps(metrics))
+
+
+   if __name__ == "__main__":
+       repo_path = Path(__file__).parent.parent
+       main(repo_path)
+   EOT
+
+Let's save the script.
+Because we're "ready for analysis" with this last piece, we can set a tag::
+
+   datalad save -m "Add script to evaluate model performance" --version-tag "ready4analysis" code/evaluate.py
+
+
+And now, we can execute the scripts.
+First, train the model::
+
+   datalad containers-run -n software \
+     -m "Train an SGD classifier on the data" \
+     --input 'data/raw/imagenette2-160/train/*' \
+     --output 'model.joblib' \
+     "python3 code/train.py"
+
+
+Then, evaluate performance::
+
+   datalad containers-run -n software \
+     -m "Evaluate SGD classifier on test data" \
+     --input 'data/raw/imagenette2-160/val/*' \
+     --output 'accuracy.json' \
+     "python3 code/evaluate.py"
+
```

Comment:
At this point I have:

```
add(ok): /tmp/ml/ml-project/code/__pycache__/train.cpython-37.pyc (file)
```

`__pycache__` should be in `.gitignore`

Reply:
~~Also `model.joblib` get's annexed. Maybe this should be treated as a temp-file and ignored too? Didn't check the code of needs to be saved....~~